File: scripts/linear_mixed_model.R
# ------------------------------------------------------------------
# Linear Mixed Model Script (nlme backend)
# ------------------------------------------------------------------
# This script is called from Python via subprocess.
# It fits a linear mixed model and writes results to CSV files.
#
# Expected arguments (in order):
# 1. input_file        : path to input CSV data
# 2. fixed_form        : fixed-effects formula (e.g., "y ~ x1 + x2")
# 3. group_var         : grouping variable (e.g., subject ID)
# 4. random_form       : random-effects formula (e.g., "1 + time")
# 5. random_covariance  : covariance structure for random effects ("diagonal", "full")
# 6. residual_covariance        : covariance structure ("independence", "ar1", "cs")
# 7. time_var          : time variable (required for AR1)
# 8. estimation_method : "REML" or "ML"
# 9. out_dir           : output directory for results
# ------------------------------------------------------------------

library(nlme)

# ------------------------------------------------------------------
# Parse command line arguments
# ------------------------------------------------------------------
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 9) {
  stop("Expected 8 arguments but received fewer.")
}

input_file              <- args[1]
fixed_form              <- args[2]
group_var               <- args[3]
random_form             <- args[4]
random_covariance       <- args[5]
residual_covariance     <- tolower(args[6])
time_var                <- args[7]
estimation_method       <- args[8]
out_dir                 <- args[9]

# ------------------------------------------------------------------
# Load data
# ------------------------------------------------------------------
df <- read.csv(input_file)

if (!(group_var %in% colnames(df))) {
  stop(paste("Grouping variable not found in data:", group_var))
}

# ------------------------------------------------------------------
# Define model components
# ------------------------------------------------------------------

# Fixed effects
fixed_formula <- as.formula(fixed_form)

# Random effects
if (random_covariance == "diagonal") {

  random_formula <- list(
    pdDiag(
      as.formula(
        paste("~", random_form)
      )
    )
  )

  names(random_formula) <- group_var

} else if (random_covariance == "full") {

  random_formula <- as.formula(
    paste("~", random_form, "|", group_var)
  )

} else {

  stop(
    paste(
      "Unsupported random covariance:",
      random_covariance
    )
  )
}

# ------------------------------------------------------------------
# Correlation structure
# ------------------------------------------------------------------
cor_struct <- NULL

if (residual_covariance == "ar1") {

  # AR(1) requires a time variable
  if (!(time_var %in% colnames(df))) {
    stop(paste("Time variable required for AR1 but not found:", time_var))
  }

  cor_struct <- corAR1(
    form = as.formula(paste("~", time_var, "|", group_var))
  )

} else if (residual_covariance == "cs") {

  cor_struct <- corCompSymm(
    form = as.formula(paste("~ 1 |", group_var))
  )

} else if (residual_covariance != "independence") {

  stop(paste("Unsupported covariance structure:", residual_covariance))
}

# ------------------------------------------------------------------
# Fit model
# ------------------------------------------------------------------
model <- lme(
  fixed = fixed_formula,
  data = df,
  random = random_formula,
  correlation = cor_struct,
  method = estimation_method
)

# ------------------------------------------------------------------
# Extract fixed effects
# ------------------------------------------------------------------
fe <- fixed.effects(model)

fe_df <- data.frame(
  term = names(fe),
  estimate = as.numeric(fe)
)

write.csv(
  fe_df,
  file.path(out_dir, "fixed_effects.csv"),
  row.names = FALSE
)

# ------------------------------------------------------------------
# Extract random effects
# ------------------------------------------------------------------
re <- ranef(model)

re_df <- data.frame(
  group = rownames(re),
  re
)

# rename 'group' → actual group_var
colnames(re_df)[1] <- group_var

# Make column names explicit (e.g., re_intercept, re_time)
colnames(re_df)[2:ncol(re_df)] <- paste0("re_", colnames(re))

write.csv(
  re_df,
  file.path(out_dir, "random_effects.csv"),
  row.names = FALSE
)

# ------------------------------------------------------------------
# Extract model metrics
# ------------------------------------------------------------------
residual_sd <- sigma(model)
residual_variance <- residual_sd^2

metrics_df <- data.frame(
  aic = AIC(model),
  bic = BIC(model),
  log_likelihood = as.numeric(logLik(model)),
  residual_sd = residual_sd,
  residual_variance = residual_variance
)

write.csv(
  metrics_df,
  file.path(out_dir, "metrics.csv"),
  row.names = FALSE
)

File: src/models/mixed_effects/linear_mixed_model.py
import subprocess
import pandas as pd
import os
import tempfile
from src.utils import helpers
from pathlib import Path

def fit_lmm_safe(lmm_fn, lmm_config, train_df, residuals, history, lmm_metrics_history, iteration):
    """Safely fit LMM and fall back to previous random effects if fitting fails."""

    df = train_df.copy()
    df["residual"] = residuals  # Add residuals as target

    try:
        outputs = lmm_fn(
            data=df,
            **lmm_config
        )
        return outputs["random_effects"], outputs["metrics"]  # Return estimated random effects

    except RuntimeError as e:
        print(f"LMM failed at iter {iteration}: {e}")  # Log failure

        if history:
            print("→ Reusing previous random effects")
            return history[-1], lmm_metrics_history[-1]  # Fallback to last successful estimate

        return None, None  # No fallback available

def fit_lmm(
    script_path: str,
    data: pd.DataFrame,
    fixed_formula: str,
    group_var: str,
    random_formula: str = "1",
    random_effects_covariance: str = "diagonal",
    residual_covariance: str = "independence",
    time_var: str = "time",
    estimation_method: str = "REML",
) -> dict:
    """
    Fit a linear mixed model (LMM) using an R backend (nlme).

    This function interfaces with an R script via subprocess,
    enabling advanced mixed-effects modeling while maintaining
    a Python-based workflow.

    Parameters
    ----------
    data : pd.DataFrame
        Input dataset containing all variables used in the model.
    fixed_formula : str
        Fixed-effects formula in R syntax (e.g., "y ~ x1 + x2").
    group_var : str
        Grouping variable for random effects (e.g., subject ID).
    random_formula : str, optional
        Random-effects structure in R syntax (default: "1",
        i.e., random intercept only).
    cov_structure : str, optional
        Covariance structure for within-group errors.
        Supported options include:
        - "independence"
        - "AR1" (autoregressive)
        - "CS" (compound symmetry)
    time_var : str, optional
        Time variable required for time-dependent covariance
        structures such as AR(1).
    estimation_method : str, optional
        Estimation method used by the model:
        - "REML" (default)
        - "ML"

    Returns
    -------
    dict
        Dictionary containing:
        - fixed_effects : pd.DataFrame
        - random_effects : pd.DataFrame
            Indexed by group variable.
        - metrics : dict
            Model fit statistics.

    Raises
    ------
    RuntimeError
        If the R script execution fails.
    FileNotFoundError
        If the R script or expected output files are missing.
    ValueError
        If input data is invalid.

    Notes
    -----
    - Uses a temporary directory to exchange data with R.
    - Requires R and the `nlme` package to be installed.
    - The R script must output:
        * fixed_effects.csv
        * random_effects.csv
        * metrics.csv
    - Error messages from R are propagated for easier debugging.
    """

    if data is None or data.empty:
        raise ValueError("Input `data` must be a non-empty DataFrame.")  # Validate input

    PROJECT_ROOT = helpers.find_project_root(Path(__file__))  # Locate project root
    script_path = os.path.join(PROJECT_ROOT, script_path)     # Resolve full script path

    if not os.path.exists(script_path):
        raise FileNotFoundError(f"R script not found: {script_path}")  # Validate script

    with tempfile.TemporaryDirectory() as tmpdir:
        input_path = os.path.join(tmpdir, "input.csv")

        # ------------------------------------------------------------------
        # Write input data
        # ------------------------------------------------------------------
        data.to_csv(input_path, index=False)  # Save data for R

        # ------------------------------------------------------------------
        # Call R script
        # ------------------------------------------------------------------
        result = subprocess.run(
            [
                "Rscript",
                script_path,
                input_path,
                fixed_formula,
                group_var,
                random_formula,
                random_effects_covariance,
                residual_covariance,
                time_var,
                estimation_method,
                tmpdir,
            ],
            capture_output=True,
            text=True,
        )

        # ------------------------------------------------------------------
        # Handle R errors explicitly
        # ------------------------------------------------------------------
        if result.returncode != 0:
            raise RuntimeError(f"R script failed:\n{result.stderr}")  # Propagate R error

        # ------------------------------------------------------------------
        # Load outputs
        # ------------------------------------------------------------------
        fixed_path = os.path.join(tmpdir, "fixed_effects.csv")
        random_path = os.path.join(tmpdir, "random_effects.csv")
        metrics_path = os.path.join(tmpdir, "metrics.csv")

        if not all(os.path.exists(p) for p in [fixed_path, random_path, metrics_path]):
            raise FileNotFoundError(
                "One or more expected output files were not generated by the R script."
            )  # Ensure outputs exist

        fixed = pd.read_csv(fixed_path)    # Load fixed effects
        random = pd.read_csv(random_path)  # Load random effects
        metrics = pd.read_csv(metrics_path)  # Load metrics

        return {
            "fixed_effects": fixed,
            "random_effects": random.set_index(group_var),  # Index by group
            "metrics": metrics.to_dict(orient="records")[0],  # Convert to dict
        }